src/array_api_stubs/_draft/_types.py

```diff
@@ -60,6 +68,949 @@ def __len__(self, /) -> int:
+    ) -> array:
+        ...
+
+    @staticmethod
+    def nonzero(x: array, /) -> tuple[array, ...]:
+        ...
+
+    @staticmethod
+    def where(condition: array, x1: array, x2: array, /) -> array:
+        ...
+
+
+class _SetFuncs(Protocol):
+    @staticmethod
+    def unique_all(x: array, /) -> tuple[array, array, array, array]:
+        ...
+
+    @staticmethod
+    def unique_counts(x: array, /) -> tuple[array, array]:
+        ...
+
+    @staticmethod
+    def unique_inverse(x: array, /) -> tuple[array, array]:
+        ...
+
+    @staticmethod
+    def unique_values(x: array, /) -> array:
+        ...
+
+
+class _SortingFuncs(Protocol):
+    @staticmethod
+    def argsort(
+        x: array, /, *, axis: int = -1, descending: bool = False, stable: bool = True
+    ) -> array:
+        ...
+
+    @staticmethod
+    def sort(
+        x: array, /, *, axis: int = -1, descending: bool = False, stable: bool = True
+    ) -> array:
+        ...
+
+
+class _StatisticalFuncs(Protocol):
+    @staticmethod
+    def max(
+        x: array,
+        /,
+        *,
+        axis: int | tuple[int, ...] | None = None,
+        keepdims: bool = False,
+    ) -> array:
+        ...
+
+    @staticmethod
+    def mean(
+        x: array,
+        /,
+        *,
+        axis: int | tuple[int, ...] | None = None,
+        keepdims: bool = False,
+    ) -> array:
+        ...
+
+    @staticmethod
+    def min(
+        x: array,
+        /,
+        *,
+        axis: int | tuple[int, ...] | None = None,
+        keepdims: bool = False,
+    ) -> array:
+        ...
+
+    @staticmethod
+    def prod(
+        x: array,
+        /,
+        *,
+        axis: int | tuple[int, ...] | None = None,
+        dtype: dtype | None = None,
+        keepdims: bool = False,
+    ) -> array:
+        ...
+
+    @staticmethod
+    def std(
+        x: array,
+        /,
+        *,
+        axis: int | tuple[int, ...] | None = None,
+        correction: int | float = 0.0,
+        keepdims: bool = False,
+    ) -> array:
+        ...
+
+    @staticmethod
+    def sum(
+        x: array,
+        /,
+        *,
+        axis: int | tuple[int, ...] | None = None,
+        dtype: dtype | None = None,
+        keepdims: bool = False,
+    ) -> array:
+        ...
+
+    @staticmethod
+    def var(
+        x: array,
+        /,
+        *,
+        axis: int | tuple[int, ...] | None = None,
+        correction: int | float = 0.0,
+        keepdims: bool = False,
+    ) -> array:
+        ...
+
+
+class _UtilityFuncs(Protocol):
+    @staticmethod
+    def all(
+        x: array,
+        /,
+        *,
+        axis: int | tuple[int, ...] | None = None,
+        keepdims: bool = False,
+    ) -> array:
+        ...
+
+    @staticmethod
+    def any(
+        x: array,
+        /,
+        *,
+        axis: int | tuple[int, ...] | None = None,
+        keepdims: bool = False,
+    ) -> array:
+        ...
+
+
+class ArrayAPILinAlgNamespace(Protocol[array]):
```

Comment:
I guess making these namespace protocols generic allows downstream users to subscript their own type hints for array class. Excuse my ignorance here—if one were not to subscript the type hint, would say `mypy` type check with our array `Protocol`?

Reply:
The advantage of making ``ArrayAPINamespace`` (and related) generic wrt the array type is for 

```python
class Array:
    def __array_namespace__() -> ArrayAPINamespace[Self]: ...
```

Now a static type checker knows that for an array-api implementing library *X*, the namespace returned by ``<X>.__array_namespace__`` work on *X* arrays.